File: app/src/main/java/com/example/msimple/adapter/BaseAdapter.java
package com.example.msimple.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;

import com.example.msimple.view.holders.BaseHolder;

import java.util.ArrayList;
import java.util.List;

public abstract class BaseAdapter<T extends BaseHolder<W>,W> extends RecyclerView.Adapter<T> {

    private List<W> mData;

    public BaseAdapter() {
        mData = new ArrayList<>();
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    @Override
    public void onBindViewHolder(@NonNull T holder, int position) {
        holder.bind(mData.get(position));

    }

    public void setNewData(List<W> newData) {
        mData = newData;
        notifyDataSetChanged();

    }

    public void appendNewData(List<W> newData) {
        mData.addAll(newData);
        notifyDataSetChanged();
    }

    public W getItemAt(int position) {
        if (position < mData.size() - 1)
            return mData.get(position);

        return null;
    }

    public List<W> getItems() {
        if (mData == null)
            return new ArrayList<>();

        return mData;
    }

    public void removeData(W data) {
        mData.remove(data);
        notifyDataSetChanged();
    }

    public void addNewData(W data) {
        mData.add(data);
        notifyDataSetChanged();
    }

    public void clearData() {
        mData = new ArrayList<>();
        notifyDataSetChanged();
    }

}

File: app/src/main/java/com/example/msimple/adapter/BaseAdapter.kt
package com.example.msimple.adapter

import android.support.v7.widget.RecyclerView

import com.example.msimple.view.holders.BaseHolder

import java.util.ArrayList

abstract class BaseAdapter<T : BaseHolder<W>, W> : RecyclerView.Adapter<T>() {

    private var mData: MutableList<W>? = null

    val items: List<W>
        get() = if (mData == null) ArrayList() else mData

    init {
        mData = ArrayList()
    }

    override fun getItemCount(): Int {
        return mData!!.size
    }

    override fun onBindViewHolder(holder: T, position: Int) {
        holder.bind(mData!![position])

    }

    fun setNewData(newData: MutableList<W>) {
        mData = newData
        notifyDataSetChanged()

    }

    fun appendNewData(newData: List<W>) {
        mData!!.addAll(newData)
        notifyDataSetChanged()
    }

    fun getItemAt(position: Int): W? {
        return if (position < mData!!.size - 1) mData!![position] else null

    }

    fun removeData(data: W) {
        mData!!.remove(data)
        notifyDataSetChanged()
    }

    fun addNewData(data: W) {
        mData!!.add(data)
        notifyDataSetChanged()
    }

    fun clearData() {
        mData = ArrayList()
        notifyDataSetChanged()
    }

}
